File: bubbler_app/lib/data/repositories/auth_repository.dart
import '../../core/api/api_client.dart';
import '../../core/api/endpoints.dart';

/// Login / register response from `/auth/*` (Swift `AuthResponse`).
class AuthResponse {
  const AuthResponse({
    required this.accessToken,
    required this.tokenType,
    required this.userId,
  });

  factory AuthResponse.fromJson(Map<String, dynamic> json) {
    return AuthResponse(
      accessToken: json['access_token'] as String,
      tokenType: json['token_type'] as String,
      userId: json['user_id'] as int,
    );
  }

  final String accessToken;
  final String tokenType;
  final int userId;
}

/// Auth domain verbs formerly on Swift `APIClient` (`login` / `register`).
class AuthRepository {
  AuthRepository(this._client);

  final ApiClient _client;

  /// OAuth2 password form: field `username` is the account email.
  Future<AuthResponse> login({
    required String email,
    required String password,
  }) async {
    final json = await _client.postForm(
      Endpoints.authLogin,
      {
        'username': email,
        'password': password,
      },
    );
    return AuthResponse.fromJson(json);
  }

  /// Registers a user. [dateOfBirth] is encoded as a local calendar day
  /// (`YYYY-MM-DD`), not a UTC instant, matching the iOS client.
  Future<AuthResponse> register({
    required String username,
    required String email,
    required String password,
    required DateTime dateOfBirth,
  }) async {
    final json = await _client.postJson(
      Endpoints.authRegister,
      {
        'username': username,
        'email': email,
        'password': password,
        'date_of_birth': formatDateOfBirth(dateOfBirth),
      },
    );
    return AuthResponse.fromJson(json);
  }

  /// Local calendar day string — avoids DOB shifting across time zones.
  static String formatDateOfBirth(DateTime dateOfBirth) {
    final local = dateOfBirth.toLocal();
    final year = local.year.toString().padLeft(4, '0');
    final month = local.month.toString().padLeft(2, '0');
    final day = local.day.toString().padLeft(2, '0');
    return '$year-$month-$day';
  }
}

File: BubblerAndroid/app/src/main/java/com/bubbler/android/data/repository/AuthRepository.kt
package com.bubbler.android.data.repository

import com.bubbler.android.core.network.ApiClient
import com.bubbler.android.core.network.Endpoints
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import okhttp3.FormBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.time.LocalDate

@Serializable
data class AuthResponse(
    @SerialName("access_token")
    val accessToken: String,
    @SerialName("token_type")
    val tokenType: String,
    @SerialName("user_id")
    val userId: Int,
)

/**
 * Auth domain HTTP calls (login / register from iOS APIClient).
 */
open class AuthRepository(
    private val apiClient: ApiClient,
) {
    /**
     * OAuth2 password form: field `username` carries the email address.
     */
    open suspend fun login(email: String, password: String): AuthResponse {
        val body = FormBody.Builder()
            .add("username", email)
            .add("password", password)
            .build()

        return apiClient.request(
            path = Endpoints.AUTH_LOGIN,
            method = "POST",
            body = body,
        )
    }

    /**
     * @param dateOfBirth calendar day only — sent as `YYYY-MM-DD`, never a UTC instant.
     */
    open suspend fun register(
        username: String,
        email: String,
        password: String,
        dateOfBirth: LocalDate,
    ): AuthResponse {
        val payload = apiClient.json.encodeToString(
            RegisterBody.serializer(),
            RegisterBody(
                username = username,
                email = email,
                password = password,
                dateOfBirth = dateOfBirth.toString(),
            ),
        )
        val body = payload.toRequestBody(ApiClient.JSON_MEDIA_TYPE)

        return apiClient.request(
            path = Endpoints.AUTH_REGISTER,
            method = "POST",
            body = body,
            contentType = "application/json",
        )
    }
}

@Serializable
private data class RegisterBody(
    val username: String,
    val email: String,
    val password: String,
    @SerialName("date_of_birth")
    val dateOfBirth: String,
)
